### PR TITLE
Add the ability to set raw private/public keys in EVP

### PR DESCRIPTION
**Pull request by mattcaswell**
The EVP interface that we have for X25519/X448/Ed25519/X25519 is very bare bones and provides just enough to use it for TLS. It is difficult to use directly because it is hard to setup private/public keys. You either have to load them from a file, or generate completely new ones. If you have the raw key data from some external source you cannot get OpenSSL to use it.

This PR adds functions to enable that. It provides a new function which is almost identical in behaviour to the existing EVP_PKEY_new_mac_key(). This PR introduces the function EVP_PKEY_new_private_key() which can be used by all the 25519/448 functions, as well as all of the functions supported by EVP_PKEY_new_mac_key(). Although the external behaviour between the two is similar the internal implementation is completely different. The initial attempt to implement these new functions was in #5419 and was heavily based on the EVP_PKEY_new_mac_key() implementation. However that implementation turned out to be controversial and was therefore removed from #5419.

This new PR implements the same functions as were removed from #5419 but does so in a completely different (and I hope less controversial) way.

I think this should be included in the beta1 release.

##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

**Comment by mattcaswell** [comment]
@kroeckx - any thoughts on this?

**Comment by levitte** [comment]
I've some thoughts on this...

1.  `EVP_PKEY_new_private_key` is a very generic name, and intuitively, I'm sure someone would want to create a RSA EVP_PKEY using this, is there any plans to expand this functionality to cover *all* public key algos that we implement?
2.  What, exactly, is "raw key data"?  Is it formatted in any way?  I'm thinking that for an RSA key, there are a few numbers, so the "raw key data" would have to be formatted precisely.  Me, I'd expect DER formatting, and quite frankly, since we often talk about raw DER (as opposed to PEM formatted DER), I can imagine others will think along the same lines.  This may be important to know for anyone wanting to implement the method functions to support this new functionality, and should be documented.

File: crypto/include/internal/asn1_int.h [hunk 1]
```diff
@@ -58,6 +58,9 @@ struct evp_pkey_asn1_method_st {
     int (*pkey_check) (const EVP_PKEY *pk);
     int (*pkey_public_check) (const EVP_PKEY *pk);
     int (*pkey_param_check) (const EVP_PKEY *pk);
+    /* Get/set raw private/public key data */
+    int (*set_priv_key) (EVP_PKEY *pk, const unsigned char *priv, size_t len);
+    int (*set_pub_key) (EVP_PKEY *pk, const unsigned char *pub, size_t len);
```
Comment (levitte) [inline crypto/include/internal/asn1_int.h]: You should add functions to set these, in `crypto/asn1/ameth_lib.c`, and document them in `doc/man3/EVP_PKEY_ASN1_METHOD.pod`

Comment (mattcaswell) [reply crypto/include/internal/asn1_int.h]: Good point. Done.

**Comment by mattcaswell** [comment]
> 1. EVP_PKEY_new_private_key is a very generic name, and intuitively, I'm sure someone would want to create a RSA EVP_PKEY using this, is there any plans to expand this functionality to cover all public key algos that we implement?
> 2. What, exactly, is "raw key data"? Is it formatted in any way? I'm thinking that for an RSA key, there are a few numbers, so the "raw key data" would have to be formatted precisely. Me, I'd expect DER formatting, and quite frankly, since we often talk about raw DER (as opposed to PEM formatted DER), I can imagine others will think along the same lines. This may be important to know for anyone wanting to implement the method functions to support this new functionality, and should be documented.

Well I was thinking slightly differently. I see the private/public keys set via these functions as just a block of binary data without any real structure to it. Instead, where there are algos which require a structured key we should have an individual function for each of them, e.g. `EVP_PKEY_new_RSA_key()`, `EVP_PKEY_new_EC_key()`, `EVP_PKEY_new_DSA_key(`).

Each of those would then take the relevant component parts required to construct the key as arguments e.g.
````
EVP_PKEY *EVP_PKEY_new_RSA_key(ENGINE *e, BIGNUM *n, BIGNUM *e, BIGNUM *d);
````
You can of course already do this today by creating an `RSA` structure first and then assigning it to the `EVP_PKEY` - but that requires use of the low level algo functions which we are trying to discourage.

But all of that is for a different PR. I wanted to see how this one was received first.

**Comment by mattcaswell** [comment]
New commit pushed addressing comments.

**Comment by mattcaswell** [comment]
Ping @levitte.

**Comment by mattcaswell** [comment]
Rebased to resolve a conflict with master. @levitte?

**Comment by levitte** [comment]
I read your explanation, and while I understand what you say and can get along with that, I stil think that a very generic name like `EVP_PKEY_new_private_key` intuitively leads people to think that it should be usable for all sorts of `EVP_PKEY`.  I can certainly go along with the thought that you simply feed it random bits and have the underlying method functions deal with it, and I can see this extend to having those functions decide the format, which could be whatever.

However, unless the intent is to have the function being used as generically as the name suggests (regardless of the itty gritty details in the docs), I cannot see that function name as acceptable.

**Comment by t-j-h** [comment]
So would EVP_PKEY_new_binary_key or EVP_PKEY_new_raw_key be better names?

**Comment by levitte** [comment]
That's not really the point...  but you know, whatever @mattcaswell says, I'm sure that sooner or later, someone is going to want to create RSA keys that way, and will implement the necessary method.  With that line of thought, I can play :-)

**Comment by levitte** [comment]
... although, even though add `raw` into the name doesn't make it much less generic, it does add a bit more clarity to what we're dealing with, so yes, I think I'd like that name change (and preferably `raw`...  if nothing else then because it's shorter ;-))

**Comment by mattcaswell** [comment]
Ok I did the rename (pushed). Is this better @levitte?

**Comment by mattcaswell** [comment]
Pushed (I fixed a trivial whitespace issue in libcrypto.num first).

**Comment by kareldonk** [comment]
Any chance to get a set of functions to export the raw key data to a buffer to use elsewhere outside of openssl?

**Comment by richsalz** [comment]
It's too late to add this new feature for this release, sorry.

**Comment by mattcaswell** [comment]
> Any chance to get a set of functions to export the raw key data to a buffer to use elsewhere outside of openssl?

Darn...I should have added that. Please could you create an issue for this?

**Comment by kareldonk** [comment]
Yes I will thanks, and even if it doesn't make it in the current release, it would be very useful to have in a future release.

**Comment by ms-s** [comment]
@mattcaswell @kareldonk I can confirm that a feature to export raw keys would be very useful. I use Curve25519 and send the public keys on the wire in JSON Web Key format (RFC7517). I can import the peer key received on the wire with EVP_PKEY_new_public_key() and derive the shared secret. For exporting I just read the last 32 bytes (BITSTRING) from the EVP_PKEY. Would be nice to have a cleaner solution in the next release.

**Pull request closed without merging**